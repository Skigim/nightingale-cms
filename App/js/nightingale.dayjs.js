/**
 * Nightingale Application Suite - Shared Day.js Utility Wrapper
 *
 * This file provides a standardized, globally accessible 'dateUtils' object
 * that wraps the Day.js library. This ensures consistent date and time
 * formatting and manipulation across all applications in the suite.
 */

// The Day.js library and its plugins are expected to be loaded globally before this script.
const dateUtils = {
  /**
   * Returns the current timestamp as an ISO 8601 string.
   * @returns {string} The current date and time in ISO format.
   */
  now: () => dayjs().toISOString(),

  /**
   * Formats a date string into a specified format.
   * @param {string} dateString - The date string to format.
   * @param {string} formatStr - The desired output format (e.g., "MM/DD/YYYY").
   * @returns {string} The formatted date string, or "N/A" if invalid.
   */
  format: (dateString, formatStr = "MM/DD/YYYY") => {
    if (!dateString) return "N/A";
    const date = dayjs(dateString);
    return date.isValid() ? date.format(formatStr) : "N/A";
  },

  /**
   * Compares two dates for sorting. Returns -1 if A is before B, 1 if A is after B, 0 if equal.
   * Handles invalid dates gracefully.
   * @param {string} dateA - The first date string.
   * @param {string} dateB - The second date string.
   * @returns {number} The comparison result (-1, 0, or 1).
   */
  compareDates: (dateA, dateB) => {
    const d1 = dayjs(dateA);
    const d2 = dayjs(dateB);
    if (!d1.isValid() || !d2.isValid()) return 0;
    if (d1.isBefore(d2)) return -1;
    if (d1.isAfter(d2)) return 1;
    return 0;
  },

  /**
   * Checks if the first date is before the second date.
   * @param {string} dateToCheck The date to evaluate.
   * @param {string} dateToCompareAgainst The baseline date.
   * @returns {boolean}
   */
  isBefore: (dateToCheck, dateToCompareAgainst) => {
    const d1 = dayjs(dateToCheck);
    const d2 = dayjs(dateToCompareAgainst);
    return d1.isValid() && d2.isValid() ? d1.isBefore(d2) : false;
  },

  /**
   * Calculates a date that was a specified number of months ago from now.
   * @param {number} months The number of months to subtract.
   * @returns {string} The resulting date in ISO format.
   */
  monthsAgo: (months) => dayjs().subtract(months, 'month').toISOString(),
};