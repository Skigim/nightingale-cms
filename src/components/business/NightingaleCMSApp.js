/**
 * NightingaleCMSApp.js - Main application component
 *
 * Root business component for the Nightingale CMS application.
 * Manages global state, services, and orchestrates all other components.
 *
 * @namespace NightingaleBusiness
 * @version 1.0.0
 * @author Nightingale CMS Team
 */

/**
 * NightingaleCMSApp Component
 * Main application component with global state management
 *
 * @returns {React.Element} NightingaleCMSApp component
 */
function NightingaleCMSApp() {
  const e = window.React?.createElement;
  const { useState, useEffect, useRef, useMemo } = window.React || {};

  // Main application state - hooks must be called unconditionally
  const [fullData, setFullData] = useState(null);
  const [activeTab, setActiveTab] = useState('dashboard');
  const [fileStatus, setFileStatus] = useState('disconnected');
  const [isSettingsOpen, setIsSettingsOpen] = useState(false);
  const [caseViewMode, setCaseViewMode] = useState('list'); // Track if we're in case details view
  const [caseBackFunction, setCaseBackFunction] = useState(null); // Function to go back from case details
  // TODO: Implement people and organizations navigation in sidebar
  // const [peopleViewMode, setPeopleViewMode] = useState('list'); // Track if we're in person details view
  // const [peopleBackFunction, setPeopleBackFunction] = useState(null); // Function to go back from person details
  // const [organizationsViewMode, setOrganizationsViewMode] = useState('list'); // Track if we're in organization details view
  // const [organizationsBackFunction, setOrganizationsBackFunction] = useState(null); // Function to go back from organization details

  // Autosave & File service state (combined)
  const [autosaveFileService, setAutosaveFileService] = useState(null);
  const [autosaveStatus, setAutosaveStatus] = useState({
    status: 'disconnected',
    message: 'Service not initialized',
  });

  // Use ref for live data access to avoid stale closure issues
  const fullDataRef = useRef(null);

  // Keep ref synchronized with state
  useEffect(() => {
    fullDataRef.current = fullData;
  }, [fullData]);

  // Backward compatibility aliases
  const fileService = autosaveFileService; // For components that expect fileService
  const [servicesReady, setServicesReady] = useState(false);

  // Get component dependencies
  const Sidebar =
    window.Sidebar || window.NightingaleUI?.getComponent?.('Sidebar');
  const Header =
    window.Header || window.NightingaleUI?.getComponent?.('Header');
  const SettingsModal =
    window.SettingsModal ||
    window.NightingaleBusiness?.getComponent?.('SettingsModal');
  const showToast = useMemo(() => {
    return (
      window.showToast ||
      window.NightingaleToast?.show ||
      function (message, type) {
        console.log(`Toast ${type}: ${message}`);
      }
    );
  }, []);

  // Initialize combined autosave & file service when services are ready
  useEffect(() => {
    const handleServicesReady = () => {
      console.log('🎯 Services ready - initializing AutosaveFileService');
      if (typeof window.AutosaveFileService !== 'undefined') {
        // Use the clean factory method for React integration
        const service = window.AutosaveFileService.createForReact({
          tabId: `cms-react-tab-${Date.now()}`,
          errorCallback: showToast,
          sanitizeFn:
            typeof window.sanitize !== 'undefined'
              ? window.sanitize
              : (str) => str,

          // Autosave configuration
          enabled: true,
          saveInterval: 120000, // 2 minutes
          debounceDelay: 5000, // 5 seconds
          maxRetries: 3,

          // React integration - use ref for live data access
          getFullData: () => {
            console.log(
              '📦 Data provider called, ref contains:',
              !!fullDataRef.current
            );
            return fullDataRef.current;
          },

          // Status updates
          onStatusChange: (status) => {
            console.log('🔄 Autosave status change:', status);
            setAutosaveStatus(status);
          },
        });

        console.log('✅ AutosaveFileService initialized:', !!service);
        setAutosaveFileService(service);
        setServicesReady(true);
      } else {
        console.warn('⚠️ AutosaveFileService not available, retrying...');
        setTimeout(handleServicesReady, 1000);
      }
    };

    // Listen for services ready event
    window.addEventListener('nightingale:services:ready', handleServicesReady);

    // Also check if already ready
    if (typeof window.AutosaveFileService !== 'undefined') {
      handleServicesReady();
    }

    return () => {
      window.removeEventListener(
        'nightingale:services:ready',
        handleServicesReady
      );
    };
  }, [showToast]);

  // Handle manual save
  const handleManualSave = async () => {
    if (fileService?.saveData) {
      try {
        await fileService.saveData();
        showToast('Data saved successfully', 'success');
      } catch (error) {
        console.error('Manual save error:', error);
        showToast('Failed to save data', 'error');
      }
    }
  };

  // Handle data updates from children
  const handleDataLoaded = (data) => {
    console.log('📊 Data loaded from child component');
    setFullData(data);
  };

  const handleDataUpdate = (newData) => {
    console.log('📊 Data updated from child component');
    setFullData(newData);
  };

  // Render tab content based on active tab
  const renderTabContent = () => {
    switch (activeTab) {
      case 'dashboard':
        return window.DashboardTab
          ? e(window.DashboardTab, { fullData })
          : e(
              'div',
              { className: 'p-4 text-center text-gray-400' },
              'Dashboard component not loaded'
            );
      case 'cases':
        return window.CasesTab
          ? e(window.CasesTab, {
              fullData,
              onUpdateData: handleDataUpdate,
              fileService,
              onViewModeChange: setCaseViewMode,
              onBackToList: setCaseBackFunction,
            })
          : e(
              'div',
              { className: 'p-4 text-center text-gray-400' },
              'Cases component not loaded'
            );
      case 'people':
        return window.PeopleTab
          ? e(window.PeopleTab, {
              fullData,
              onUpdateData: handleDataUpdate,
              fileService,
              // TODO: Implement navigation when sidebar supports people back button
              // onViewModeChange: setPeopleViewMode,
              // onBackToList: setPeopleBackFunction,
            })
          : e(
              'div',
              { className: 'p-4 text-center text-gray-400' },
              'People component not loaded'
            );
      case 'organizations':
        return window.OrganizationsTab
          ? e(window.OrganizationsTab, {
              fullData,
              onUpdateData: handleDataUpdate,
              fileService,
              // TODO: Implement navigation when sidebar supports organizations back button
              // onViewModeChange: setOrganizationsViewMode,
              // onBackToList: setOrganizationsBackFunction,
            })
          : e(
              'div',
              { className: 'p-4 text-center text-gray-400' },
              'Organizations component not loaded'
            );
      case 'eligibility':
        return window.EligibilityTab
          ? e(window.EligibilityTab, { fullData })
          : e(
              'div',
              { className: 'p-4 text-center text-gray-400' },
              'Eligibility component not loaded'
            );
      default:
        return window.DashboardTab
          ? e(window.DashboardTab, { fullData })
          : e(
              'div',
              { className: 'p-4 text-center text-gray-400' },
              'Dashboard component not loaded'
            );
    }
  };

  // React safety check for render
  if (!window.React) {
    console.warn('React not available for NightingaleCMSApp component');
    return null;
  }

  // Show loading screen while services are initializing
  if (!servicesReady || !fileService) {
    return e(
      'div',
      {
        className:
          'h-screen w-screen flex items-center justify-center bg-gray-900 text-white',
        style: { fontFamily: 'Inter, system-ui, sans-serif' },
      },
      e(
        'div',
        { className: 'text-center space-y-4' },
        e('div', {
          className:
            'animate-spin h-8 w-8 border-4 border-blue-500 border-t-transparent rounded-full mx-auto',
        }),
        e(
          'h2',
          { className: 'text-xl font-semibold' },
          'Initializing Nightingale CMS'
        ),
        e(
          'p',
          { className: 'text-gray-400' },
          'Loading services and file system...'
        )
      )
    );
  }

  return e(
    'div',
    { className: 'h-screen w-screen flex' },
    Sidebar &&
      e(Sidebar, {
        activeTab,
        onTabChange: setActiveTab,
        onSettingsClick: () => setIsSettingsOpen(true),
        caseViewMode,
        onCaseBackToList: () => {
          if (caseBackFunction) {
            caseBackFunction();
          }
        },
      }),
    e(
      'div',
      { className: 'flex-1 flex flex-col overflow-hidden' },
      Header &&
        e(Header, {
          fileStatus,
          autosaveStatus,
          onSettingsClick: () => setIsSettingsOpen(true),
          onManualSave: handleManualSave,
        }),
      e(
        'main',
        { className: 'flex-1 overflow-auto p-6 bg-gray-900' },
        renderTabContent()
      )
    ),
    // Settings Modal
    SettingsModal &&
      e(SettingsModal, {
        isOpen: isSettingsOpen,
        onClose: () => setIsSettingsOpen(false),
        fileService,
        onDataLoaded: handleDataLoaded,
        fileStatus,
        onFileStatusChange: setFileStatus,
      })
  );
}

// PropTypes for validation
if (typeof window !== 'undefined' && window.PropTypes) {
  NightingaleCMSApp.propTypes = {
    // NightingaleCMSApp is the root component and doesn't receive external props
    // All props are managed internally through state
  };
}

// Self-registration for both module and script loading
if (typeof window !== 'undefined') {
  // Register globally for backward compatibility
  window.NightingaleCMSApp = NightingaleCMSApp;

  // Register with NightingaleBusiness registry if available
  if (window.NightingaleBusiness) {
    window.NightingaleBusiness.registerComponent(
      'NightingaleCMSApp',
      NightingaleCMSApp,
      'application'
    );
  }
}

// Export for ES6 module compatibility
if (typeof module !== 'undefined' && module.exports) {
  module.exports = NightingaleCMSApp;
}
