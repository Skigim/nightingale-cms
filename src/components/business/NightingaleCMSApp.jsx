/**
 * NightingaleCMSApp.js - Main application component
 *
 * Root business component for the Nightingale CMS application.
 * Manages global state, services, and orchestrates all other components.
 *
 * @version 1.0.0
 * @author Nightingale CMS Team
 */

// Direct ES module imports for primary tab components (eliminates reliance on global registry lookups)
import React, {
  useState,
  useEffect,
  useRef,
  useMemo,
  useCallback,
} from 'react';
import DashboardTab from './DashboardTab.jsx';
import CasesTab from './CasesTab.jsx';
import PeopleTab from './PeopleTab.jsx';
import OrganizationsTab from './OrganizationsTab.jsx';
import EligibilityTab from './EligibilityTab.jsx';
import { ThemeProvider, createTheme } from '@mui/material/styles';
import CssBaseline from '@mui/material/CssBaseline';
import { registerComponent, getComponent } from '../../services/registry';
import Toast from '../../services/nightingale.toast.js';
import { getFileService } from '../../services/fileServiceProvider.js';
// Keep Header / Sidebar / SettingsModal via global for now (can be migrated later)

/**
 * NightingaleCMSApp Component
 * Main application component with global state management
 *
 * @returns {React.Element} NightingaleCMSApp component
 */
function NightingaleCMSApp() {
  // Main application state - hooks must be called unconditionally
  const [fullData, setFullData] = useState(null);
  const [activeTab, setActiveTab] = useState('dashboard');
  const [fileStatus, setFileStatus] = useState('disconnected');
  const [isSettingsOpen, setIsSettingsOpen] = useState(false);
  const [caseViewMode, setCaseViewMode] = useState('list'); // Track if we're in case details view
  const [caseBackFunction, setCaseBackFunction] = useState(null); // Function to go back from case details
  const [isBugModalOpen, setIsBugModalOpen] = useState(false);
  // TODO: Implement people and organizations navigation in sidebar
  // const [peopleViewMode, setPeopleViewMode] = useState('list'); // Track if we're in person details view
  // const [peopleBackFunction, setPeopleBackFunction] = useState(null); // Function to go back from person details
  // const [organizationsViewMode, setOrganizationsViewMode] = useState('list'); // Track if we're in organization details view
  // const [organizationsBackFunction, setOrganizationsBackFunction] = useState(null); // Function to go back from organization details

  // File service instance (hydrated from provider)
  const [fileService, setFileServiceState] = useState(() => getFileService());
  const [autosaveStatus, setAutosaveStatus] = useState(null);

  // Keep AutosaveFileService synced with current data and status callback
  useEffect(() => {
    if (!fileService) return;
    try {
      if (typeof fileService.initializeWithReactState === 'function') {
        fileService.initializeWithReactState(
          () => fullDataRef.current,
          (status) => {
            setAutosaveStatus(status);
            if (status?.status === 'waiting') {
              setFileStatus('reconnect');
            }
          },
        );
      } else if (typeof fileService.setDataProvider === 'function') {
        // Fallback: at least ensure provider is set
        fileService.setDataProvider(() => fullDataRef.current);
      }
    } catch (_) {
      // no-op; UI will show disconnected
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [fileService]);

  // Use ref for live data access to avoid stale closure issues
  const fullDataRef = useRef(null);

  // Data update handler - must be defined before useMemo that uses it
  const handleDataUpdate = useCallback(
    (newData) => {
      setFullData(newData);
      // Notify autosave service debounced save may proceed
      try {
        fileService?.notifyDataChange?.();
      } catch (_) {
        /* ignore */
      }
    },
    [fileService],
  );

  // Keep ref synchronized with state
  useEffect(() => {
    fullDataRef.current = fullData;
  }, [fullData]);

  // Listen for provider readiness events to hydrate service after mount
  useEffect(() => {
    const handler = () => {
      const svc = getFileService();
      if (svc && svc !== fileService) {
        setFileServiceState(svc);
      }
    };
    // Attempt immediate hydration (handles case where service set before mount)
    handler();
    if (typeof document !== 'undefined') {
      document.addEventListener('fileService:ready', handler);
    }
    return () => {
      if (typeof document !== 'undefined') {
        document.removeEventListener('fileService:ready', handler);
      }
    };
    // fileService intentionally excluded from deps to avoid stale removal/add cycle
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // When fileService changes, reflect connection status and auto-load data
  useEffect(() => {
    let cancelled = false;
    (async () => {
      if (!fileService?.checkPermission) return;
      try {
        const perm = await fileService.checkPermission();
        if (cancelled) return;
        if (perm === 'granted') {
          setFileStatus('connected');
          if (typeof fileService.readFile === 'function') {
            const data = await fileService.readFile();
            if (cancelled) return;
            if (data && Object.keys(data).length > 0) {
              setFullData(data);
            }
          }
        } else {
          setFileStatus('reconnect');
        }
      } catch (_) {
        setFileStatus('disconnected');
      }
    })();
    return () => {
      cancelled = true;
    };
  }, [fileService]);

  // Memoize non-tab components (tabs imported directly above)
  const components = useMemo(
    () => ({
      Sidebar: getComponent('ui', 'Sidebar'),
      Header: getComponent('ui', 'Header'),
      SettingsModal: getComponent('business', 'SettingsModal'),
      BugReportModal: getComponent('business', 'BugReportModal'),
    }),
    [],
  );

  // Memoize tab props to prevent unnecessary re-renders of heavy tab components
  const tabProps = useMemo(
    () => ({
      dashboard: { fullData },
      cases: {
        fullData,
        onUpdateData: handleDataUpdate,
        fileService,
        onViewModeChange: setCaseViewMode,
        onBackToList: setCaseBackFunction,
      },
      people: {
        fullData,
        onUpdateData: handleDataUpdate,
        fileService,
        // TODO: Add navigation props when sidebar supports people back button
        // onViewModeChange: setPeopleViewMode,
        // onBackToList: setPeopleBackFunction,
      },
      organizations: {
        fullData,
        onUpdateData: handleDataUpdate,
        fileService,
        // TODO: Add navigation props when sidebar supports organizations back button
        // onViewModeChange: setOrganizationsViewMode,
        // onBackToList: setOrganizationsBackFunction,
      },
      eligibility: { fullData },
    }),
    [
      fullData,
      fileService,
      handleDataUpdate,
      setCaseViewMode,
      setCaseBackFunction,
    ],
  );

  // Memoize service status to optimize conditional rendering
  const serviceStatus = useMemo(
    () => ({
      isReady: !!fileService,
      hasAutosave: !!fileService, // unified service now
      canSave: !!fileService?.save,
      isLoading: false,
    }),
    [fileService],
  );

  // Toast function - module based
  const showToast = (msg, type) => Toast.showToast?.(msg, type);

  const handleSubmitBug = useCallback(
    async ({ content, activeTab: tab, createdAt }) => {
      try {
        const logger = globalThis.NightingaleLogger?.get('bug-report');
        logger?.info('Bug report submitted', { tab, createdAt, content });
        const entry = {
          id: `${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
          tab,
          content,
          createdAt,
        };
        // Persist to localStorage list first (source of truth for aggregation)
        const key = 'nightingale-bug-reports';
        const list = JSON.parse(localStorage.getItem(key) || '[]');
        list.push(entry);
        localStorage.setItem(key, JSON.stringify(list));

        // Attempt to mirror to project folder as JSON array
        if (fileService?.writeNamedFile) {
          const fileName = 'bug-reports.json';
          await fileService.writeNamedFile(fileName, list);
          showToast('Bug report saved to project folder', 'success');
        } else {
          showToast('Bug report stored locally (no file access)', 'warning');
        }
      } catch (err) {
        const logger = globalThis.NightingaleLogger?.get('bug-report');
        logger?.error('Failed to submit bug report', { error: err.message });
        showToast('Failed to save bug report', 'error');
      } finally {
        setIsBugModalOpen(false);
      }
    },
    [fileService],
  );

  // (Legacy services ready listener removed) Autosave/file service will now be
  // initialized elsewhere; component renders immediately.

  // Handle manual save
  const handleManualSave = async () => {
    if (serviceStatus.canSave) {
      try {
        await fileService.save();
        showToast('Data saved successfully', 'success');
      } catch (error) {
        const logger = globalThis.NightingaleLogger?.get('cms:manualSave');
        logger?.error('Manual save failed', { error: error.message });
        showToast('Failed to save data', 'error');
      }
    } else {
      showToast('Save service not available', 'warning');
    }
  };

  // Handle data updates from children
  const handleDataLoaded = useCallback(
    (data) => {
      setFullData(data);
      try {
        fileService?.notifyDataChange?.();
      } catch (_) {
        /* ignore */
      }
    },
    [fileService],
  );

  // Render active tab using lookup mapping (JSX friendly)
  const renderActiveTab = () => {
    const tabComponents = {
      dashboard: DashboardTab,
      cases: CasesTab,
      people: PeopleTab,
      organizations: OrganizationsTab,
      eligibility: EligibilityTab,
    };
    const TabComponent = tabComponents[activeTab];
    const props = tabProps[activeTab];
    if (TabComponent && props) {
      // Using React.createElement to avoid potential JSX parse edge in dynamic component
      return React.createElement(TabComponent, { ...props });
    }
    return React.createElement(
      'div',
      { className: 'p-4 text-center text-gray-400' },
      `${activeTab.charAt(0).toUpperCase() + activeTab.slice(1)} component not loaded`,
    );
  };

  // React is imported via ESM; no global safety check needed

  // Global MUI dark theme aligned with Tailwind colors used in the app
  const theme = React.useMemo(
    () =>
      createTheme({
        palette: {
          mode: 'dark',
          primary: { main: '#3B82F6' }, // blue-500
          secondary: { main: '#6366F1' }, // indigo-500
          background: {
            default: '#111827', // gray-900 (matches app main background)
            paper: '#1F2937', // gray-800
          },
          text: {
            primary: '#E5E7EB', // gray-200
            secondary: '#9CA3AF', // gray-400
          },
        },
        components: {
          MuiCssBaseline: {
            styleOverrides: {
              // Theme the scrollbars for the main app content area only
              'main.flex-1': {
                scrollbarWidth: 'thin',
                scrollbarColor: '#4B5563 #111827', // thumb track (Firefox)
              },
              'main.flex-1::-webkit-scrollbar': {
                width: '10px',
                height: '10px',
              },
              'main.flex-1::-webkit-scrollbar-track': {
                background: '#111827', // gray-900
                borderRadius: 8,
              },
              'main.flex-1::-webkit-scrollbar-thumb': {
                backgroundColor: '#4B5563', // gray-600
                borderRadius: 8,
                border: '2px solid #111827',
              },
              'main.flex-1::-webkit-scrollbar-thumb:hover': {
                backgroundColor: '#6B7280', // gray-500
              },
            },
          },
          MuiButton: {
            styleOverrides: {
              root: { textTransform: 'none', borderRadius: 8 },
            },
          },
          MuiPaper: {
            styleOverrides: {
              root: { backgroundImage: 'none' },
            },
          },
          MuiTableCell: {
            styleOverrides: {
              head: { fontWeight: 600 },
            },
          },
          MuiDataGrid: {
            styleOverrides: {
              root: {
                // Customize native scrollbars for WebKit and Firefox
                '& .MuiDataGrid-virtualScroller': {
                  scrollbarWidth: 'thin', // Firefox
                  scrollbarColor: '#4B5563 #111827', // thumb track
                },
                '& .MuiDataGrid-virtualScroller::-webkit-scrollbar': {
                  width: '10px',
                  height: '10px',
                },
                '& .MuiDataGrid-virtualScroller::-webkit-scrollbar-track': {
                  background: '#111827', // bg-gray-900
                  borderRadius: 8,
                },
                '& .MuiDataGrid-virtualScroller::-webkit-scrollbar-thumb': {
                  backgroundColor: '#4B5563', // gray-600
                  borderRadius: 8,
                  border: '2px solid #111827',
                },
                '& .MuiDataGrid-virtualScroller::-webkit-scrollbar-thumb:hover':
                  {
                    backgroundColor: '#6B7280', // gray-500
                  },
              },
            },
          },
        },
      }),
    [],
  );

  return React.createElement(
    ThemeProvider,
    { theme },
    React.createElement(CssBaseline, null),
    React.createElement(
      'div',
      { className: 'h-screen w-screen flex' },
      components.Sidebar &&
        React.createElement(components.Sidebar, {
          activeTab,
          onTabChange: setActiveTab,
          onSettingsClick: () => setIsSettingsOpen(true),
          onReportBugClick: () => setIsBugModalOpen(true),
          caseViewMode,
          onCaseBackToList: () => {
            if (caseBackFunction) caseBackFunction();
          },
        }),
      React.createElement(
        'div',
        { className: 'flex-1 flex flex-col overflow-hidden' },
        components.Header &&
          React.createElement(components.Header, {
            fileStatus,
            autosaveStatus,
            onSettingsClick: () => setIsSettingsOpen(true),
            onManualSave: handleManualSave,
          }),
        React.createElement(
          'main',
          { className: 'flex-1 overflow-auto p-6 bg-gray-900' },
          renderActiveTab(),
        ),
      ),
      components.SettingsModal &&
        React.createElement(components.SettingsModal, {
          isOpen: isSettingsOpen,
          onClose: () => setIsSettingsOpen(false),
          fileService,
          onDataLoaded: handleDataLoaded,
          fileStatus,
          onFileStatusChange: setFileStatus,
        }),
      components.BugReportModal &&
        React.createElement(components.BugReportModal, {
          isOpen: isBugModalOpen,
          onClose: () => setIsBugModalOpen(false),
          onSubmit: handleSubmitBug,
          activeTab,
        }),
    ),
  );
}

// Self-registration for both module and script loading
// Register with business registry (legacy global removal)
registerComponent('business', 'NightingaleCMSApp', NightingaleCMSApp);

export default NightingaleCMSApp;
