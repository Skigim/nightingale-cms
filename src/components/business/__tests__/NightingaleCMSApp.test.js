/**
 * @jest-environment jsdom
 */

import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import React from 'react';

// Import the component
require('../NightingaleCMSApp.js');

describe('NightingaleCMSApp Component', () => {
  beforeEach(() => {
    // Ensure React is available globally
    global.window = window;
    window.React = React;

    // Mock required services and components
    window.NightingaleServices = {
      loadData: jest.fn().mockResolvedValue({
        cases: [],
        people: [],
        organizations: [],
      }),
      saveData: jest.fn().mockResolvedValue(),
      createFileService: jest.fn().mockReturnValue({
        save: jest.fn(),
        load: jest.fn(),
        getFileInfo: jest.fn().mockReturnValue({ name: 'test.json' }),
      }),
    };

    // Mock AutosaveFileService
    window.AutosaveFileService = {
      createForReact: jest.fn().mockReturnValue({
        on: jest.fn(),
        off: jest.fn(),
        save: jest.fn(),
        load: jest.fn(),
      }),
    };

    // Mock component dependencies
    window.Header = jest.fn(() => React.createElement('div', null, 'Header'));
    window.Sidebar = jest.fn(() => React.createElement('div', null, 'Sidebar'));
    window.ErrorBoundary = jest.fn(({ children }) => children);
    window.DashboardTab = jest.fn(() =>
      React.createElement('div', null, 'Dashboard')
    );
    window.CasesTab = jest.fn(() => React.createElement('div', null, 'Cases'));
    window.PeopleTab = jest.fn(() =>
      React.createElement('div', null, 'People')
    );
    window.OrganizationsTab = jest.fn(() =>
      React.createElement('div', null, 'Organizations')
    );
    window.EligibilityTab = jest.fn(() =>
      React.createElement('div', null, 'Eligibility')
    );
    window.SettingsModal = jest.fn(() =>
      React.createElement('div', null, 'Settings')
    );

    // Mock toast service
    window.NightingaleToast = {
      show: jest.fn(),
    };

    jest.clearAllMocks();
  });

  test('renders without crashing', () => {
    const Component = window.NightingaleCMSApp;
    expect(Component).toBeDefined();

    render(React.createElement(Component));
  });

  test('initializes autosave service correctly', async () => {
    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Component should initialize AutosaveFileService
    await waitFor(
      () => {
        expect(window.AutosaveFileService.createForReact).toHaveBeenCalled();
      },
      { timeout: 2000 }
    );
  });

  test('loads data on mount', async () => {
    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Wait for component to initialize dashboard tab (which shows data is ready)
    await waitFor(() => {
      expect(window.DashboardTab).toHaveBeenCalled();
    });
  });

  test('renders header component with correct props', () => {
    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Header should be rendered with required props (based on actual implementation)
    expect(window.Header).toHaveBeenCalledWith(
      expect.objectContaining({
        fileStatus: expect.any(String),
        autosaveStatus: expect.any(Object),
        onManualSave: expect.any(Function),
        onSettingsClick: expect.any(Function),
      }),
      {}
    );
  });

  test('renders sidebar component', () => {
    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    expect(window.Sidebar).toHaveBeenCalledWith(
      expect.objectContaining({
        activeTab: expect.any(String),
        onTabChange: expect.any(Function),
      }),
      expect.any(Object)
    );
  });

  test('changes active tab when sidebar tab is clicked', async () => {
    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Get the onTabChange callback from Sidebar props
    const sidebarProps = window.Sidebar.mock.calls[0][0];
    const onTabChange = sidebarProps.onTabChange;

    // Simulate tab change
    onTabChange('cases');

    await waitFor(() => {
      expect(window.CasesTab).toHaveBeenCalled();
    });
  });

  test('opens settings modal when header settings is clicked', async () => {
    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Get the onSettingsClick callback from Header props
    const headerProps = window.Header.mock.calls[0][0];
    const onSettingsClick = headerProps.onSettingsClick;

    // Simulate settings click
    onSettingsClick();

    await waitFor(() => {
      expect(window.SettingsModal).toHaveBeenCalledWith(
        expect.objectContaining({
          isOpen: true,
        }),
        expect.any(Object)
      );
    });
  });

  test('handles manual save from header', async () => {
    // Mock autosave service with saveData method
    const mockSaveData = jest.fn().mockResolvedValue();
    window.AutosaveFileService.createForReact.mockReturnValue({
      on: jest.fn(),
      off: jest.fn(),
      save: jest.fn(),
      load: jest.fn(),
      saveData: mockSaveData,
    });

    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Wait for autosave service initialization
    await waitFor(() => {
      expect(window.AutosaveFileService.createForReact).toHaveBeenCalled();
    });

    // Get the onManualSave callback from Header props
    const headerCall =
      window.Header.mock.calls[window.Header.mock.calls.length - 1];
    const headerProps = headerCall[0];
    const onManualSave = headerProps.onManualSave;

    // Simulate manual save
    await onManualSave();

    // Should use autosave service for saving
    expect(mockSaveData).toHaveBeenCalled();
  });

  test('provides correct initial status to header', async () => {
    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Check initial header props
    const headerProps = window.Header.mock.calls[0][0];
    expect(headerProps.fileStatus).toBe('disconnected');
    expect(headerProps.autosaveStatus).toEqual({
      message: 'Service not initialized',
      status: 'disconnected',
    });
  });

  test('handles autosave status updates', async () => {
    // Mock autosave service that calls status callback
    let statusCallback;
    window.AutosaveFileService.createForReact.mockImplementation((config) => {
      statusCallback = config.onStatusChange;
      return {
        on: jest.fn(),
        off: jest.fn(),
        save: jest.fn(),
        load: jest.fn(),
      };
    });

    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Wait for autosave service initialization
    await waitFor(() => {
      expect(window.AutosaveFileService.createForReact).toHaveBeenCalled();
    });

    // Simulate status change from autosave service
    if (statusCallback) {
      statusCallback({
        status: 'connected',
        message: 'Ready',
      });
    }

    // Header should receive updated status
    await waitFor(() => {
      const headerCall =
        window.Header.mock.calls[window.Header.mock.calls.length - 1];
      const headerProps = headerCall[0];
      expect(headerProps.autosaveStatus.status).toBe('connected');
    });
  });

  test('tracks dirty state correctly', async () => {
    // Mock autosave service that calls dirty state callback
    let dirtyCallback;
    window.AutosaveFileService.createForReact.mockImplementation((config) => {
      dirtyCallback = config.onDirtyChange;
      return {
        on: jest.fn(),
        off: jest.fn(),
        save: jest.fn(),
        load: jest.fn(),
      };
    });

    const Component = window.NightingaleCMSApp;

    render(React.createElement(Component));

    // Wait for autosave service initialization
    await waitFor(() => {
      expect(window.AutosaveFileService.createForReact).toHaveBeenCalled();
    });

    // Simulate dirty state change from autosave service
    if (dirtyCallback) {
      dirtyCallback(true);
    }

    // Component should track dirty state but doesn't pass it to Header in current implementation
    // This test verifies the callback is set up correctly
    expect(dirtyCallback).toBeDefined();
  });

  test('handles component unmounting gracefully', () => {
    const Component = window.NightingaleCMSApp;

    const { unmount } = render(React.createElement(Component));

    expect(() => unmount()).not.toThrow();
  });

  test('matches snapshot', async () => {
    const Component = window.NightingaleCMSApp;

    const { container } = render(React.createElement(Component));

    // Wait for dashboard tab to render (indicates component is initialized)
    await waitFor(() => {
      expect(window.DashboardTab).toHaveBeenCalled();
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
