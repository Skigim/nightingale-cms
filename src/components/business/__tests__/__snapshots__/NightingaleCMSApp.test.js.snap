// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NightingaleCMSApp Component matches snapshot 1`] = `
<div
  class="h-screen w-screen flex"
  style=""
>
  <div>
    Sidebar
  </div>
  <div
    class="flex-1 flex flex-col overflow-hidden"
  >
    <div>
      Header
    </div>
    <main
      class="flex-1 overflow-auto p-6 bg-gray-900"
    >
      <div>
        Dashboard
      </div>
    </main>
  </div>
  <div>
    Settings
  </div>
</div>
`;
